File: .agents/teamwork_preview_explorer_m2_3/extract_lessons.py
import os
import re
import json

migrations_dir = r"C:\Users\UTHtest\.gemini\antigravity\worktrees\network\fix-lesson-completion-logic\supabase\migrations"
seed_path = r"C:\Users\UTHtest\.gemini\antigravity\worktrees\network\fix-lesson-completion-logic\src\app\data\seed.sql"

lessons = {}

# Helper to normalize and add/update a lesson
def update_lesson(lesson_id, title_en=None, content_en=None):
    if lesson_id not in lessons:
        lessons[lesson_id] = {"id": lesson_id, "title_en": None, "content_en": None}
    if title_en:
        lessons[lesson_id]["title_en"] = title_en.strip()
    if content_en:
        # Resolve dollar quotes if present
        if content_en.startswith("$$") and content_en.endswith("$$"):
            content_en = content_en[2:-2]
        lessons[lesson_id]["content_en"] = content_en.strip()

# 1. Parse seed.sql
print("Parsing seed.sql...")
if os.path.exists(seed_path):
    with open(seed_path, "r", encoding="utf-8") as f:
        content = f.read()
    
    # Parse INSERT INTO lessons
    # Pattern to find the INSERT statement values block
    insert_match = re.search(r"INSERT INTO lessons \([^)]*\) VALUES\s*(.*?);", content, re.DOTALL | re.IGNORECASE)
    if insert_match:
        values_block = insert_match.group(1)
        # Parse individual tuples: ('id', 'course_id', 'title_th', 'title_en', 'content_th', 'content_en', ...)
        # We can extract strings within single quotes, paying attention to escaped quotes
        # Since it's a regular structure: ('id','course','title_th','title_en','content_th','content_en',...)
        # We can split by values pattern
        tuples = re.findall(r"\(([^)]+)\)", values_block)
        for t in tuples:
            # simple parse by comma-splitting after resolving quote groupings
            # Since it's a seed file, let's use a regex to match single quoted strings
            parts = re.findall(r"'((?:[^'\\]|\\.)*)'", t)
            if len(parts) >= 6:
                lid = parts[0]
                title_en = parts[3]
                content_en = parts[5]
                update_lesson(lid, title_en=title_en, content_en=content_en)
                # print(f"Seed lesson found: {lid} - {title_en}")

# 2. Parse all migrations
print("Parsing migrations...")
sql_files = sorted([f for f in os.listdir(migrations_dir) if f.endswith(".sql")])

for file_name in sql_files:
    file_path = os.path.join(migrations_dir, file_name)
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    
    # Check for INSERT INTO public.lessons
    inserts = re.findall(r"INSERT INTO public\.lessons \([^)]*\) VALUES\s*\(([^)]+)\);", content, re.DOTALL | re.IGNORECASE)
    for ins in inserts:
        parts = re.findall(r"'((?:[^'\\]|\\.)*)'", ins)
        if len(parts) >= 6:
            lid = parts[0]
            title_en = parts[3]
            content_en = parts[5]
            update_lesson(lid, title_en=title_en, content_en=content_en)
            # print(f"Migration insert found in {file_name}: {lid} - {title_en}")

    # Check for UPDATE public.lessons or UPDATE lessons
    # Pattern 1: UPDATE public.lessons SET content_en = '...' WHERE id = '...';
    # Pattern 2: UPDATE public.lessons SET content_th = '...', content_en = '...' WHERE id = '...';
    # Or setting content_en with dollar-quotes: content_en = $$...$$
    
    # We can search for all UPDATE statements targeting lessons
    updates = re.finditer(r"UPDATE\s+(?:public\.)?lessons\s+SET\s+(.*?)\s+WHERE\s+id\s*=\s*'([^']+)';", content, re.DOTALL | re.IGNORECASE)
    for match in updates:
        set_clause = match.group(1)
        lid = match.group(2)
        
        # Try to extract content_en
        # It could be single quoted or dollar quoted
        content_en_match = re.search(r"content_en\s*=\s*('\s*(?:[^'\\]|\\.|'')*'\s*|\$\$(.*?)\$\$)", set_clause, re.DOTALL | re.IGNORECASE)
        title_en_match = re.search(r"title_en\s*=\s*('\s*(?:[^'\\]|\\.|'')*'\s*|\$\$(.*?)\$\$)", set_clause, re.DOTALL | re.IGNORECASE)
        
        title_en = None
        content_en = None
        
        if title_en_match:
            val = title_en_match.group(1)
            if val.startswith("$$"):
                title_en = title_en_match.group(2)
            else:
                title_en = val[1:-1].replace("''", "'")
                
        if content_en_match:
            val = content_en_match.group(1)
            if val.startswith("$$"):
                content_en = content_en_match.group(2)
            else:
                content_en = val[1:-1].replace("''", "'")
                
        if title_en or content_en:
            update_lesson(lid, title_en=title_en, content_en=content_en)
            # print(f"Migration update found in {file_name}: {lid}")

# 3. Save to a file for analysis
output_path = r"C:\Users\UTHtest\.gemini\antigravity\worktrees\network\fix-lesson-completion-logic\.agents\teamwork_preview_explorer_m2_3\lessons_extracted.json"
with open(output_path, "w", encoding="utf-8") as f:
    json.dump(list(lessons.values()), f, indent=2, ensure_ascii=False)

print(f"Extraction completed. Total lessons found: {len(lessons)}")
print(f"Saved to: {output_path}")

File: .agents/teamwork_preview_explorer_m2_3/extract_lessons.js
import fs from 'fs';
import path from 'path';

const migrationsDir = 'C:\\Users\\UTHtest\\.gemini\\antigravity\\worktrees\\network\\fix-lesson-completion-logic\\supabase\\migrations';
const seedPath = 'C:\\Users\\UTHtest\\.gemini\\antigravity\\worktrees\\network\\fix-lesson-completion-logic\\src\\app\\data\\seed.sql';
const outputPath = 'C:\\Users\\UTHtest\\.gemini\\antigravity\\worktrees\\network\\fix-lesson-completion-logic\\.agents\\teamwork_preview_explorer_m2_3\\lessons_extracted.json';

const lessons = {};

function updateLesson(id, title_en, content_en) {
  if (!id) return;
  if (!lessons[id]) {
    lessons[id] = { id, title_en: null, content_en: null };
  }
  if (title_en) {
    lessons[id].title_en = title_en.trim();
  }
  if (content_en) {
    // Strip dollar quotes if present
    let cleaned = content_en.trim();
    if (cleaned.startsWith('$$') && cleaned.endsWith('$$')) {
      cleaned = cleaned.substring(2, cleaned.length - 2);
    }
    lessons[id].content_en = cleaned.trim();
  }
}

// 1. Parse seed.sql
console.log('Parsing seed.sql...');
if (fs.existsSync(seedPath)) {
  const content = fs.readFileSync(seedPath, 'utf8');
  
  // Find INSERT INTO lessons VALUES block
  const insertRegex = /INSERT\s+INTO\s+lessons\s*\([^)]*\)\s*VALUES\s*([\s\S]*?);/i;
  const match = content.match(insertRegex);
  if (match) {
    const valuesBlock = match[1];
    // Find each tuple block like ('lesson-id', ...)
    const tupleRegex = /\(([^)]+)\)/g;
    let m;
    while ((m = tupleRegex.exec(valuesBlock)) !== null) {
      const tupleContent = m[1];
      // Match single-quoted strings: '...' or double-single quotes inside
      // A simple regex to split single-quoted elements
      const parts = [];
      const partRegex = /'((?:[^'\\]|\\.|'')*)'/g;
      let pm;
      while ((pm = partRegex.exec(tupleContent)) !== null) {
        parts.push(pm[1].replace(/''/g, "'"));
      }
      if (parts.length >= 6) {
        const id = parts[0];
        const title_en = parts[3];
        const content_en = parts[5];
        updateLesson(id, title_en, content_en);
      }
    }
  }
}

// 2. Parse migrations
console.log('Parsing migrations...');
const files = fs.readdirSync(migrationsDir).filter(f => f.endsWith('.sql')).sort();

for (const file of files) {
  const filePath = path.join(migrationsDir, file);
  const content = fs.readFileSync(filePath, 'utf8');

  // Parse INSERT statements
  const insertRegex = /INSERT\s+INTO\s+public\.lessons\s*\([^)]*\)\s*VALUES\s*\(([\s\S]*?)\);/gi;
  let im;
  while ((im = insertRegex.exec(content)) !== null) {
    const tupleContent = im[1];
    const parts = [];
    const partRegex = /'((?:[^'\\]|\\.|'')*)'/g;
    let pm;
    while ((pm = partRegex.exec(tupleContent)) !== null) {
      parts.push(pm[1].replace(/''/g, "'"));
    }
    if (parts.length >= 6) {
      const id = parts[0];
      const title_en = parts[3];
      const content_en = parts[5];
      updateLesson(id, title_en, content_en);
    }
  }

  // Parse UPDATE statements targeting public.lessons or lessons
  // E.g., UPDATE public.lessons SET content_en = '...' WHERE id = '...';
  // Or UPDATE public.lessons SET content_th = '...', content_en = '...' WHERE id = '...';
  // We can use a regex for: UPDATE [public.]lessons SET [clauses] WHERE id = '...';
  // Let's capture the set clause and the id
  const updateRegex = /UPDATE\s+(?:public\.)?lessons\s+SET\s+([\s\S]*?)\s+WHERE\s+id\s*=\s*'([^']+)';/gi;
  let um;
  while ((um = updateRegex.exec(content)) !== null) {
    const setClause = um[1];
    const id = um[2];

    // Try to extract content_en
    // It can be single-quoted '...' or dollar-quoted $$...$$
    let content_en = null;
    const contentEnMatch = setClause.match(/content_en\s*=\s*(?:'((?:[^'\\]|\\.|'')*)'|\$\$([\s\S]*?)\$\$)/i);
    if (contentEnMatch) {
      content_en = contentEnMatch[1] ? contentEnMatch[1].replace(/''/g, "'") : '$$' + contentEnMatch[2] + '$$';
    }

    // Try to extract title_en
    let title_en = null;
    const titleEnMatch = setClause.match(/title_en\s*=\s*(?:'((?:[^'\\]|\\.|'')*)'|\$\$([\s\S]*?)\$\$)/i);
    if (titleEnMatch) {
      title_en = titleEnMatch[1] ? titleEnMatch[1].replace(/''/g, "'") : '$$' + titleEnMatch[2] + '$$';
    }

    if (title_en || content_en) {
      updateLesson(id, title_en, content_en);
    }
  }
}

// 3. Write output
const lessonList = Object.values(lessons);
fs.writeFileSync(outputPath, JSON.stringify(lessonList, null, 2), 'utf8');
console.log(`Saved ${lessonList.length} lessons to ${outputPath}`);

// Check which lessons have empty content_en or empty title_en
const incomplete = lessonList.filter(l => !l.title_en || !l.content_en);
console.log(`Lessons missing title_en or content_en: ${incomplete.length}`);
if (incomplete.length > 0) {
  console.log('Incomplete lessons:', incomplete.map(l => l.id));
}
